engine/namegen.ml

```diff
@@ -320,6 +320,24 @@ let next_name_away_in_goal na avoid =
     | Anonymous -> default_non_dependent_ident in
   next_ident_away_in_goal id avoid
 
+(** Check whether [id] is already defined in the current namespaces,
+    i.e., in any of the enclosing sections in the current module path. *)
+let is_defined_here id =
```

Comment:
Why not use `Global.exists_objlabel`? Or move the logic in `Declare.exists_name` to here? I don't really understand the difference in behaviour though.

Reply:
Where can I find the documentation of these functions?

Reply:
>Where can I find the documentation of these functions?

You look at the implementation

Reply:
Thanks for the pointers. The following patch seems good enough to address the issue:

~~~
--- a/engine/namegen.ml
+++ b/engine/namegen.ml
@@ -329,7 +329,7 @@ let next_name_away_in_goal na avoid =
 
 let next_global_ident_away id avoid =
   let id = if Id.Set.mem id avoid then restart_subscript id else id in
-  let bad id = Id.Set.mem id avoid || is_global id in
+  let bad id = Id.Set.mem id avoid || Global.exists_objlabel (Label.of_id id) in
   next_ident_away_from id bad
 
 (* 4- Looks for next fresh name outside a list; if name already used,
~~~

Let me run the CI to see if something would break.